clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,697 @@
+
+export type JobId = string;
+export interface JobDescription {
+    jobname: string;
+    id: JobId;
+    details: string;
+}
+
+function getJobDetailInfo(job: JobDescription): { rc: number; spoolFiles: number } | undefined {
+    const parsed = /^.*RC=(\d+)\s+(\d+) spool file/.exec(job.details);
+    if (!parsed)
+        return undefined;
+    else
+        return { rc: +parsed[1], spoolFiles: +parsed[2] };
+}
+export interface JobClient {
+    submitJcl(jcl: string): Promise<JobId>;
+    setListMask(mask: string): Promise<void>;
+    list(): Promise<JobDescription[]>;
+    download(target: Writable | string, id: JobId, spoolFile: number): Promise<void>;
+    close(): void;
+}
+
+async function basicFtpJobClient(connection: {
+    host: string;
+    port?: number;
+    user: string;
+    password: string;
+    secure: connectionSecurityLevel
+}): Promise<JobClient> {
+    const client = new Client();
+    client.parseList = (rawList: string): FileInfo[] => {
+        return rawList.split(/\r?\n/).slice(1).filter(x => !/^\s*$/.test(x)).map((value) => new FileInfo(value));
+    };
+    await client.access({
+        host: connection.host,
+        user: connection.user,
+        password: connection.password,
+        port: connection.port,
+        secure: connection.secure !== connectionSecurityLevel.unsecure,
+        secureOptions: connection.secure === connectionSecurityLevel.unsecure ? undefined : { rejectUnauthorized: connection.secure !== connectionSecurityLevel.rejectUnauthorized }
+    });
+
+    const checkResponse = (resp: FTPResponse) => {
+        if (resp.code < 200 || resp.code > 299)
+            throw Error("FTP Error: " + resp.message);
+    }
+    const checkedCommand = async (command: string): Promise<string> => {
+        const resp = await client.send(command);
+        checkResponse(resp);
+        return resp.message
+    }
+    const switchText = async () => { await checkedCommand("TYPE A"); }
+    const switchBinary = async () => { await checkedCommand("TYPE I"); }
+
+    await checkedCommand("SITE FILE=JES");
+    return {
+        async submitJcl(jcl: string): Promise<string> {
+            await switchText();
+            const jobUpload = await client.uploadFrom(Readable.from(jcl), "JOB");
+            checkResponse(jobUpload);
+            const jobid = /^.*as ([Jj](?:[Oo][Bb])?\d+)/.exec(jobUpload.message);
+            if (!jobid)
+                throw Error("Unable to extract the job id");
+            return jobid[1];
+        },
+        async setListMask(mask: string): Promise<void> {
+            await checkedCommand("SITE JESJOBNAME=" + mask);
+            await checkedCommand("SITE JESSTATUS=OUTPUT");
+        },
+        async list(): Promise<JobDescription[]> {
+            try {
+                await switchText();
+                return (await client.list()).map((x: FileInfo): JobDescription => {
+                    const parsedLine = /(\S+)\s+(\S+)\s+(.*)/.exec(x.name);
```

Comment:
Should the last capturing group capture just 'non-whitespace' characters to be safe?
```suggestion
                    const parsedLine = /(\S+)\s+(\S+)\s+(\S+)/.exec(x.name);
```

Reply:
if the JCL fails at the parsing stage the rc and spoolFiles do not exist, so I think we should keep the two-stage parsing.

Reply:
That's a good point. In that case the type should be `{jobname, id, status?: {rc, spoolFiles}}`. No strong option about it. We can keep as is.